plugins: Migrate plugins to vite

sniok opened the pull request:
This PR continues vite migration. First part is in PR here https://github.com/headlamp-k8s/headlamp/pull/1981

Fixes this https://github.com/headlamp-k8s/headlamp/issues/1282

### Source Maps

A little note on source maps. The current webpack build generates eval source maps, they are usually only intended for development only. 
With vite we create source maps and append them to the end of the file.
In addition to that the headlamp eval part now includes a filename and an origin for the main.js bundled plugin which makes it much easier to work with source maps.

![image](https://github.com/user-attachments/assets/a4668681-84d4-429e-b070-15024a19945f)

Each plugin is given it's own fake origin


### Plugin Compatibility Test

#### Building and running plugins

Build, run and manually validate plugin works

| Name | Status | Comment |
| --- | --- | --- |
| app-catalog | Works | |
| backstage | Works | |
| flux | Crashes | Incorrect import and usage of Event.default.useList is causing the crash | 
| kompose | Works | |
| plugin-catalog | Works | |
| prometheus | Works | |
| edp-headlamp | Crashes | Unrelated to vite. Uses methods from notistack that was upgraded <br /> Uses old xterm that was upgraded. <br /> After runinng npm i xterm ... and removing notistack usage it works fine |
| examples/change-log/ | Works | |
| examples/cluster-chooser | Works | |
| examples/details-view | Works | |
| examples/headlamp-events | Works | |
| examples/pod-counter | Works | |
| examples/sidebar | Works | |
| examples/tables | Works | |

So for flux plugin we'll have to update the imports during upgrade to the latest `headlamp-plugin` version. I haven't seen any other case where it was imported like that.

For edp-headlamp they'll have to bundle their own xterm and notistack or adapt the code to use latest version. 

#### Running jest tests

Since this PR changes testing framework from jest to vitest we need to test it still works
Vitest looks like it's completely compatible with jest and isn't causing any issues

| Name | Status | Comment |
| --- | --- | --- |
| edp-headlamp | Works* | They have their own testing setup | 
| prometheus | Works | Jest tests work without any modifications | 

### Storybook and storyshots

Storyshots plugin has been deprecated and we can't support it anymore. The only usage I've seen is only in our plugins and it's limited. 

The fact that storybook in plugins has been broken for almost a year https://github.com/headlamp-k8s/headlamp/issues/1980 and no one seems to complain makes me think that as of now it shouldn't be a priority for us, especially storyshots.

illume reviewed (changes requested):
Can you please rebase against main?

(Also I'm not sure if that snyk check failing is still relevant? I thought you had made a change for that issue already?)

illume commented on plugins/headlamp-plugin/template/src/storybook.test.tsx:
```diff
@@ -1,5 +0,0 @@
-// @ts-nocheck
-import { initTests } from '@kinvolk/headlamp-plugin/config/storyshots/storyshots-test';
```
Are you sure this can be removed?

sniok replied:
since it's a template for new plugins so I don't see a risk there. If storyshots plugin is deprecated I think it's fair for us to not support it. Keeping storyshots tests for plugins is not trivial, we would effectively need to maintain a copy of 'storyshots-tests' plugin that works with vite

illume commented on plugins/headlamp-plugin/package.json:
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@kinvolk/headlamp-plugin",
-  "version": "0.9.2",
+  "version": "0.10.0",
```
Can you please leave the version as is? It can be incremented in a separate commit later.

illume commented on frontend/package-lock.json:
Were the changes to frontend/ a rebase mistake? If not, can you please move them to a separate commit?

sniok replied:
moved to a different commit

illume reviewed (changes requested):
Noticed a few things.

Seems the package exclude change needs to be added in here as well?

illume commented:
The issue for this is here: https://github.com/headlamp-k8s/headlamp/issues/1282 (should go in the PR description)

illume commented:
Does this allow plugins to access environment variables at build time too? Here's the issue for that:
- https://github.com/headlamp-k8s/headlamp/issues/1894

(but it shouldn't block this PR being merged if it doesn't support it yet)

sniok commented:
It does not provide `process` now and I'm not sure if it should. Plugin authors can use `import.meta.env` for env variables, and if they really need process they can install polyfill in the plugin

sniok commented:
Rebased and fixed merge conflicts

sniok commented:
Added `interop: auto` to the vite config that builds plugins

This allows plugins use default imports from pluginLib the same way webpack plugins do. It checks for the __esModule field and will get the default export. 

Tested the Kompose plugin (built with vite) with main (vite) and with 0.24.1 (webpack)

With this change all combinations of webpack frontend, vite frontend, webpack plugin, vite plugin should work together

sniok commented:
Did some additional testing
Collected some public plugins + example plugins (17 total) compiled all of the with vite and webpack separately.
Ran them on latest main and then 0.24.1 and didn't see any obvious problem or crashes. Seems like we won't have to mark them as incompatible

joaquimrocha commented:
Removing this from milestone v0.26.0.

illume commented:
Can we skip the storyshots? If no one is using it?

sniok commented:
Tested one external plugin and it's now broken because `notistack` dependency was updated :C

illume commented:
Which plugin was broken, and how?

sniok commented:
> Which plugin was broken, and how?

edp-headlamp, the epam one. There's an explanation in the PR description

illume commented:
@sniok can you please fix the conflicts?

illume commented:
Is it much effort to reenable the headlamp-plugin tests?

sniok commented:
@illume conflicts were resolved.

I haven't looked into what it would take to fix and reenable tests, that can be done in a separate pr

illume commented:
@sniok When I try to make the tar ball for headlamp-plugin with `npm run build && npm pack` I get a bunch of errors. Note, similar (maybe the same) errors are in the main branch as well.

sniok commented:
> When I try to make the tar ball for headlamp-plugin with `npm run build && npm pack` I get a bunch of errors. Note, similar (maybe the same) errors are in the main branch as well.

I've built and packed this branch on several occasions and didn't receive any errors. Make sure to run `git clean -fd` or manually remove folders in `headlamp/plugins/headlamp-plugin/src/` that are ignored by git since they may contain some old files.

illume commented:
@sniok thanks, cleaning src/ helped! 

It works with vite-plugins branch now.
(still some errors reported on main... but we can ignore that because they are fixed here).

illume commented:
In the plugins repo, I found this issue with app-catalog plugin.
```
cd plugins
npx @kinvolk/headlamp-plugin@alpha upgrade --headlamp-plugin-version=alpha app-catalog/

> app-catalog@0.6.0 tsc
> headlamp-plugin tsc

".": tsc-ing, :node_modules/.bin/tsc --noEmit:...
error TS2688: Cannot find type definition file for 'vite-plugin-svgr/client'.
  The file is in the program because:
    Entry point of type library 'vite-plugin-svgr/client' specified in compilerOptions

error TS2688: Cannot find type definition file for 'vite/client'.
  The file is in the program because:
    Entry point of type library 'vite/client' specified in compilerOptions


Found 2 errors.
```

sniok commented:
I can't reproduce that error locally. 


<details>

```
➜  plugins git:(main) npx @kinvolk/headlamp-plugin@alpha upgrade --headlamp-plugin-version=alpha app-catalog/
Need to install the following packages:
@kinvolk/headlamp-plugin@0.11.0-alpha.3
Ok to proceed? (y) y

Upgrading "app-catalog/"...
Updating file: "src/headlamp-plugin.d.ts"
Updated package.json field "overrides": {"typescript":"5.6.2"}
Resetting node_modules folder for more robust package upgrade...
Resetting package-lock.json file for more robust package upgrade...
Running cmd:"npm install @kinvolk/headlamp-plugin@alpha --save" inside of . abs: "/home/dubenko/plugins/app-catalog"

changed 1 package, and audited 1229 packages in 3s

387 packages are looking for funding
  run `npm fund` for details

5 moderate severity vulnerabilities

To address issues that do not require attention, run:
  npm audit fix

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.
Running cmd:"npm install -D @headlamp-k8s/eslint-config@latest --save" inside of . abs: "/home/dubenko/plugins/app-catalog"

added 8 packages, removed 1 package, changed 3 packages, and audited 1236 packages in 8s

387 packages are looking for funding
  run `npm fund` for details

5 moderate severity vulnerabilities

To address issues that do not require attention, run:
  npm audit fix

To address all issues (including breaking changes), run:
  npm audit fix --force

Run `npm audit` for details.
Running cmd:"npm audit fix" inside of app-catalog/ abs: "/home/dubenko/plugins/app-catalog/app-catalog"

changed 4 packages, and audited 1236 packages in 4s

387 packages are looking for funding
  run `npm fund` for details

# npm audit report

prismjs  <1.30.0
Severity: moderate
PrismJS DOM Clobbering vulnerability - https://github.com/advisories/GHSA-x7hr-w5r2-h6wg
fix available via `npm audit fix --force`
Will install react-syntax-highlighter@5.8.0, which is a breaking change
node_modules/refractor/node_modules/prismjs
  refractor  <=4.6.0
  Depends on vulnerable versions of prismjs
  node_modules/refractor
    react-syntax-highlighter  >=6.0.0
    Depends on vulnerable versions of refractor
    node_modules/react-syntax-highlighter

3 moderate severity vulnerabilities

To address all issues (including breaking changes), run:
  npm audit fix --force
Problem running npm audit fix
"npm audit fix" failed. You may need to inspect your dependencies manually.
Running cmd:"npm run format" inside of app-catalog/ abs: "/home/dubenko/plugins/app-catalog/app-catalog"

> app-catalog@0.6.0 format
> headlamp-plugin format

".": format-ing, :node_modules/.bin/prettier --config package.json --write --cache src:...
src/api/charts.tsx 191ms
src/api/releases.tsx 23ms
src/api/repository.tsx 5ms
src/components/charts/ChartsList.stories.tsx 24ms
src/components/charts/Details.tsx 32ms
src/components/charts/EditorDialog.tsx 61ms
src/components/charts/List.tsx 43ms
src/components/charts/SettingsLink.stories.tsx 3ms
src/components/charts/SettingsLink.tsx 6ms
src/components/releases/Detail.tsx 22ms
src/components/releases/EditorDialog.tsx 24ms
src/components/releases/List.tsx 6ms
src/components/releases/ReleaseList.stories.tsx 5ms
src/components/settings/AppCatalogSettings.stories.tsx 6ms
src/components/settings/AppCatalogSettings.tsx 5ms
src/components/settings/EnableSwitch.Stories.tsx 4ms
src/components/settings/EnableSwitch.tsx 5ms
src/headlamp-plugin.d.ts 3ms
src/helpers/index.tsx 5ms
src/index.tsx 6ms
src/storybook.test.tsx 2ms
Done format-ing: ".".

Running cmd:"npm run lint" inside of app-catalog/ abs: "/home/dubenko/plugins/app-catalog/app-catalog"

> app-catalog@0.6.0 lint
> headlamp-plugin lint

".": lint-ing, :node_modules/.bin/eslint --cache -c package.json --max-warnings 0 --ext .js,.ts,.tsx src/:...
Done lint-ing: ".".

Running cmd:"npm run tsc" inside of app-catalog/ abs: "/home/dubenko/plugins/app-catalog/app-catalog"

> app-catalog@0.6.0 tsc
> headlamp-plugin tsc

".": tsc-ing, :node_modules/.bin/tsc --noEmit:...
Done tsc-ing: ".".

Successfully upgraded: "app-catalog/".
```

</details>

And also it seems like it's installing latest version instead of alpha
`Running cmd:"npm install -D @headlamp-k8s/eslint-config@latest --save"` 

So I'm not sure if that's a good test for alpha release, we can retry once it's tagged as latest.

Also I'm sure that linting is working without issues because I've just fixed that in CI for all packages. 
PR https://github.com/headlamp-k8s/plugins/pull/178
Latest CI logs https://github.com/headlamp-k8s/plugins/actions/runs/13834620606/job/38706622842

illume commented:
@joaquimrocha can you please review?

Of note, this disables the storyshot tests which we are using in several of the plugins. This in effect means these tests are being disabled. So this is a small risk, but I think it's ok if we are aware of it until there is a replacement for storybook tests.

illume reviewed (approved):
🎉 thanks
